tests/unit/test_rdt.py

```diff
@@ -0,0 +1,320 @@
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+import pytest
+import tfs
+
+from optics_functions.constants import PHASE_ADV, X, Y, BETA, S, TUNE, NAME, REAL, IMAG
+from optics_functions.rdt import calculate_rdts, generator, get_all_to_order, str2jklm, jklm2str
+from optics_functions.utils import prepare_twiss_dataframe
+
+INPUT = Path(__file__).parent.parent / "inputs"
+
+
+@pytest.mark.basic
+def test_jklm_str():
+    n = 5
+    rdt_list = get_all_to_order(n)
+    rdt_list2 = [str2jklm(jklm2str(*r)) for r in rdt_list]
+    assert rdt_list == rdt_list2
+    assert (sum(str2jklm(r)) == 3 for r in ["F1002", "F1101", "F1020"])
+    assert (sum(str2jklm(r)) == 4 for r in ["F2002", "F2011", "F1201"])
+
+
+@pytest.mark.basic
+def test_get_all_to_order():
+    n = 5
+    rdt_list = get_all_to_order(n)
+    assert all(sum(jklm) <= n for jklm in rdt_list)
+    assert len(set(rdt_list)) == len(rdt_list)
+
+
+@pytest.mark.basic
+def test_generator():
+    n = 5
+    rdt_all = generator(list(range(2, n+1)))
+    rdt_skew = generator(list(range(2, n+1)), normal=False)
+    rdt_normal = generator(list(range(2, n+1)), skew=False)
+    rdt_cc = generator(list(range(2, n+1)), complex_conj=False)
+
+    for o in rdt_all.keys():
+        assert len(rdt_all[o]) == 2*len(rdt_skew[o])
+        assert len(rdt_all[o]) == 2*len(rdt_normal[o])
+        assert len(rdt_all[o]) == 2*len(rdt_cc[o])
+        assert set(rdt_all[o]) == set(rdt_skew[o] + rdt_normal[o])
+        assert all(is_odd(sum(rdt[2:4])) for rdt in rdt_skew[o])
+        assert all(is_even(sum(rdt[2:4])) for rdt in rdt_normal[o])
+        rdt_cc2 = set(rdt_all[o]) - set(rdt_cc[o])
+        assert all((r[1], r[0], r[3], r[2]) in rdt_cc[o] for r in rdt_cc2)
+
+
+@pytest.mark.basic
+def test_generator_vs_all_to_order():
+    n = 5
+    rdt_dict = generator(list(range(2, n+1)))
+    rdt_list = get_all_to_order(n)
+    rdt_list2 = []
+    for o, v in rdt_dict.items():
+        rdt_list2 += v
+
+    set_rdts = set(rdt_list)
+    set_rdts2 = set(rdt_list2)
+    assert not len(set_rdts2 - set_rdts)
+    assert not len(set_rdts - set_rdts2)
+
+
+@pytest.mark.basic
+def test_rdts_normal_sextupole_bump():
+    df = get_df(n=7)
+    df = prepare_twiss_dataframe(beam=1, df_twiss=df)
+    df.loc["3", "K2L"] = 1
+    df.loc["5", "K2L"] = -1
+
+    df_rdts = calculate_rdts(df, rdts=["F1002", "F2001"])
+    df_diff, df_jump = get_absdiff_and_jumps(df_rdts)
+    assert not df_rdts.isna().any().any()
+    assert not df_jump["F2001"].any()
+    assert df_jump.loc[["3", "5"], "F1002"].all()
+    assert not df_jump.loc[df_jump.index.difference(["3", "5"]), "F1002"].any()
+
+
+@pytest.mark.basic
+def test_wrong_rdt_names():
+    df = get_df(n=8)
+    df = prepare_twiss_dataframe(beam=1, df_twiss=df)
+    with pytest.raises(ValueError):
+        calculate_rdts(df, rdts=["F10003", "F1002"])
```

Comment:
there are RDT with 5 digits when including offmom, see https://ados.web.psi.ch/slsnotes/sls0997.pdf

Reply:
Not in these calculations.

Reply:
Also there are RDTs with 6 digits, if you want to cover the full dimensionality of it.

Reply:
see http://mad.web.cern.ch/mad/webguide/manual.html#Ch32.S3 (the lines directly above this)

Reply:
then maybe to be 100% exact (and keep the option for including them at some point) a NotImplementedError if either 5 or 6 indices are provided could be added ... but not a showstopper, just if easily fixable and no side effects

Reply:
I do not like the idea of having a different Exception for 5-6 length wrong RDTs as for any other wrong RDT. Makes it harder to `try: except`, especially as both are equally invalid for the formulas at hand. There is not intention to implement 5 and 6 indice RDTs

Reply:
Just comment: one can go to more than 6-index RDTs, i.e. separate indices for free and driven motion. But it is not yet output from PTC.